File: app/src/main/java/info/nightscout/androidaps/plugins/aps/openAPSMA/LoggerCallback.java
package info.nightscout.androidaps.plugins.aps.openAPSMA;

import org.mozilla.javascript.ScriptableObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.logging.L;

/**
 * Created by adrian on 15/10/17.
 */


public class LoggerCallback extends ScriptableObject {

    private static Logger log = LoggerFactory.getLogger(L.APS);

    private static StringBuffer errorBuffer = new StringBuffer();
    private static StringBuffer logBuffer = new StringBuffer();


    public LoggerCallback() {
        //empty constructor needed for Rhino
        errorBuffer = new StringBuffer();
        logBuffer = new StringBuffer();
    }

    @Override
    public String getClassName() {
        return "LoggerCallback";
    }

    public void jsConstructor() {
        //empty constructor on JS site; could work as setter
    }

    public void jsFunction_log(Object obj1) {
        if (L.isEnabled(L.APS))
            log.debug(obj1.toString());
        logBuffer.append(obj1.toString());
        logBuffer.append(' ');
    }

    public void jsFunction_error(Object obj1) {
        if (L.isEnabled(L.APS))
            log.error(obj1.toString());
        errorBuffer.append(obj1.toString());
        errorBuffer.append(' ');
    }


    public static String getScriptDebug() {
        String ret = "";
        if (errorBuffer.length() > 0) {
            ret += "e:\n" + errorBuffer.toString();
        }
        if (ret.length() > 0 && logBuffer.length() > 0) ret += '\n';
        if (logBuffer.length() > 0) {
            ret += "d:\n" + logBuffer.toString();
        }
        return ret;
    }
}

File: plugins/aps/src/main/java/info/nightscout/plugins/aps/logger/LoggerCallback.kt
package info.nightscout.plugins.aps.logger

import info.nightscout.plugins.aps.utils.StaticInjector
import info.nightscout.rx.logging.AAPSLogger
import info.nightscout.rx.logging.LTag
import org.mozilla.javascript.ScriptableObject
import javax.inject.Inject

@Suppress("unused", "FunctionName")
class LoggerCallback : ScriptableObject() {

    @Inject lateinit var aapsLogger: AAPSLogger

    override fun getClassName(): String = "LoggerCallback"

    fun jsConstructor() {
        //empty constructor on JS site; could work as setter
    }

    fun jsFunction_log(obj1: Any) {
        aapsLogger.debug(LTag.APS, obj1.toString().trim { it <= ' ' })
        logBuffer.append(obj1.toString())
    }

    fun jsFunction_error(obj1: Any) {
        aapsLogger.error(LTag.APS, obj1.toString().trim { it <= ' ' })
        errorBuffer.append(obj1.toString())
    }

    companion object {

        private var errorBuffer = StringBuffer()
        private var logBuffer = StringBuffer()
        val scriptDebug: String
            get() {
                var ret = ""
                if (errorBuffer.isNotEmpty()) {
                    ret += """
                    e:
                    $errorBuffer
                    """.trimIndent()
                }
                if (ret.isNotEmpty() && logBuffer.isNotEmpty()) ret += '\n'
                if (logBuffer.isNotEmpty()) {
                    ret += """
                    d:
                    $logBuffer
                    """.trimIndent()
                }
                return ret
            }
    }

    init {
        //empty constructor needed for Rhino
        errorBuffer = StringBuffer()
        logBuffer = StringBuffer()
        @Suppress("DEPRECATION")
        StaticInjector.getInstance().androidInjector().inject(this)
    }
}